### PR TITLE
Alias Feature HLD

### PR DESCRIPTION
**Pull request by justinejose91**

### CODE REVIEW

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+## 2.1 Target Deployment Use Cases
+
+## 2.2 Functional Description
+
+# 3 Design
+
+## 3.1 Overview
+
+- When alias mode is enabled, ethernet interfaces will be represented as Eth[slot/port/breakout-port], where slot will start from 1 and port/breakout-port numbers will start from 1. This name format is fixed and will not have references to interface speed. For e.g Eth1/1 (in non breakout mode), Eth1/1/1 (in breakout mode).
```
Comment (Chandra-BS) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Any specific reason for starting the port from 1, and not from 0. 
Also breakout mode. I think Eth 1/1:1 is more indicative and clear, instead of 1/1/1

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+#### Alias mode configuration
+
+`[no] alias enable`
+```
+sonic(config)# alias enable
+sonic(config)# no alias enable
+```
+#### Config commands when alias mode is enabled
+`Eth 1/2`, `Eth1/2`, and `e1/2` options are supported to get into alias interface config mode.  The parser converts all to `Eth1/2`.
```
Comment (Chandra-BS) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: This means that there is no long or short name for interface i.e Ethernet1/1 will be rejected in config commands (all ui type) and will not be part of any show commands

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+#### Show commands when alias mode is enabled
+```
+show vlan
+Q: A - Access (Untagged), T - Tagged
+NUM        Status      Q Ports
+2          Inactive    A  Eth1/3
+                       T  Eth1/5/2
+                       T  Eth1/10
+```
+
```
Comment (Chandra-BS) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: There are multiple way of representing the interface name in config mode. However, in show output (across all features), can we assume that it will always be like “Eth1/2” or Eth1/1/2

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+### 3.6.2 CLI
+
+#### 3.6.2.1 Configuration Commands
+
+#### Alias mode configuration
+
+`[no] alias enable`
+```
+sonic(config)# alias enable
```
Comment (Chandra-BS) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Generally “alias” is a command to create aliases for existing commands. So for extensibility and to avoid confusion with other vendors command, we should consider changing it to something like this.
o	[no] alias eth-intf-name enable

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+```
+sonic(config)# interface e1/2
+sonic(conf-if-Eth1/2)#
+```
+```
+sonic(conf-if-Eth1/2)# ip address 2.2.2.2/24
+```
+
+#### 3.6.2.2 Show Commands
```
Comment (Chandra-BS) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:261]: Provide a CLI show command to display the mapping between native and alias name (for all interfaces)
Native Name                           Alias name
Etherent1                                 Eth 1/1

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 2 above)
Comment (bhavini-gada) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: Short form is only when alias is enabled. Native mode will have long form. This has been discussed & agreed by both teams earlier.

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 3 above)
Comment (bhavini-gada) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: All feature owners have to modify respective outputs to be inline with this format.

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 5 above)
Comment (bhavini-gada) [reply doc/mgmt/SONiC_AliasFeature_HLD.md:261]: Yes, agree !

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 1 above)
Comment (bhavini-gada) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: The format & port numbering agreed by both teams earlier !

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+sonic(config)# alias enable
+sonic# show alias
+Alias mode is enabled
+```
+```
+sonic(config)# no alias enable
+sonic# show alias
+Alias mode is disabled
+```
+#### Show commands when alias mode is enabled
```
Comment (nnag-broadcom) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: While the show commands are displaying the ports in aliasing mode, what happens when a user disable aliasing?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+
+#### Configuration via gNMI
+
+Same as CLI configuration test, but using gNMI SET request
+
+#### Get configuration via gNMI
+
+Same as CLI show test, but using gNMI GET request, verify the JSON response.
+
+#### Configuration via REST
```
Comment (nnag-broadcom) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:401]: Can you please mention the PUT & PATCH for alias enable cli ?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+sonic(conf-if-Eth1/2)# ip address 2.2.2.2/24
+```
+
+#### 3.6.2.2 Show Commands
+
+#### Display alias mode  
+
+`show alias`
+```
+sonic(config)# alias enable
```
Comment (nnag-broadcom) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Is there a show command to check if alias mode is enable or disabled?

**Review by nnag-broadcom — CHANGES_REQUESTED** [review]
Please check the review comments

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 6 above)
Comment (justinejose91) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: Native interface name will be used if alias mode is disabled.

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 8 above)
Comment (justinejose91) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: Please have a look at section #show commands, line# 241

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 6 above)
Comment (justinejose91) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: Native interface name format is explained in Basic Approach section.

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 9]
```diff
@@ -0,0 +1,337 @@
+# Alias Feature
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: I think we should re-think the whole "alias" naming, and use another keyword (e.g. "standard"). I feel that we are overloading the use of "alias" here and this is going to create confusion. For instance: -
- In SONiC (port, breakout), alias is used to describe user-defined naming, whereas this is a fixed format
- In Cisco, an alias is a user-configured shortcut name for a more complex command
- Same in the Linux shell

Basically this is not aliasing - this is changing the interface naming mode in the SONiC Mgmt Framework.

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+|      |            |              |                    |
+|      |            |              |                    |
+
+# About this Manual
+
+This document provides general information about alias related syntax & overview for ethernet Interfaces in break out & non-breakout mode.
+
+# Scope
+
+Covers general design for supporting alias feature.
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: I think it's important to note that the scope of this is limited to the SONiC Mgmt Framework (I think!). SONiC Click is unaffected I believe (at least it's not mentioned at all). Also config_db.json - what's the story here? The scoping needs stating.

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+
+### 1.1.4 Warm Boot Requirements
+
+## 1.2 Design Overview
+
+### 1.2.1 Basic Approach
+
+- By default system will boot in native mode. Native mode is where interfaces are displayed as Ethernet 1, Ethernet 4, Ethernet 8 etc. This also ensures that backwards compatibility is maintained.
+- Once alias is enabled, all subsequent CLI, REST & gNMI interfaces will use the alternative name for ethernet interface configuration & retrieval.
+- Existing SSH sessions : TBD.
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Please address the TBD

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 1 above)
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Please define what you mean by "slot" in the SONiC world.

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+## 2.1 Target Deployment Use Cases
+
+## 2.2 Functional Description
+
+# 3 Design
+
+## 3.1 Overview
+
+- When alias mode is enabled, ethernet interfaces will be represented as Eth[slot/port/breakout-port], where slot will start from 1 and port/breakout-port numbers will start from 1. This name format is fixed and will not have references to interface speed. For e.g Eth1/1 (in non breakout mode), Eth1/1/1 (in breakout mode).
+- The alias name {e.g Eth1/1) is picked from platform.json file. All platforms to be updated with this format for port name.
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Are you referring to the "alias_at_lanes" defined here - https://github.com/Azure/SONiC/blob/439944c120d59edf7f62ed28a902b1a3c07a35f8/doc/dynamic-port-breakout/sonic-dynamic-port-breakout-HLD.md? Please say so. As I say above, I'm not convinced we should be overloading this with a fixed format.

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+  *1. Add “value-transformer” annotation for the leaves & keys that need special handling, in the sonic-yang annot file.*  
+  *2. Write value-transformer callback api*  
+
+  Actions that Xfmr-infra will do:  
+  *1. Process dbDataMap & invoke value-transformer, if present.*
+
+
+  **Use case:**
+
+“value-transformer” (callback_xfmr) annotated for a leaf(leaf-A) in a sonic-yang(assume soinc-A.yang), will be invoked for that leaf(leaf-A).
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: typo "soinc"

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+
+If leaf-A in sonic-A.yang is referred from multiple other yang-leaves(assume leaf-B in sonic-B.yang & leaf-C in sonic-C.yang has leaf-ref to leaf-A in sonic-A.yang), then the “callback_xfmr” will be inherited & invoked for these leaves(leaf-B & C) as well, without any explicit annotation(i.e. no “value-transformer” annotation is needed for leaf-B or leaf-C).
+
+  User can override inherited value-transformer, by annotating a new value-xfmr at that level. i.e. If an explicit value-transformer(“callback_NEW_xfmr”) is annotated for leaf-B in soinc-B.yang(which has leaf-ref to leaf-A), then “callback_NEW_xfmr” will be invoked for leaf-B.
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: typo - "soinc"

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+
+
+
+   ***GET operation(OC & Sonic)**
+
+| **Existing  implementation**                                 | **New implementation**                                       |
+| ------------------------------------------------------------ | ------------------------------------------------------------ |
+| For any CRU request that reaches Xfmr, Xfmr-infra will read data from db and  create dbDataMap.  It then invokes overloaded  functions(fieldxfmr, subtreexfmr…) with dbDataMap. Fills the ygot tree with  data from dbDataMap & merges it with ygot-tree from subtree-xfmr and sends response back. | For any CRU request that reaches Xfmr, xfmr-infra will read data from db and create dbDataMap. First value-xfmr (if present) is invoked, dbDataMap is processed and a new processed-dbDataMap  is created.  Overloaded  functions(fieldxfmr, subtreexfmr…) with processed-dbDataMap will be invoked next. Fills the ygot tree with  data from dbDataMap & merges it with ygot-tree from subtree-xfmr and sends response back. |
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Why are you talking about "CRU" operations in a section on "Get operation"?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+   Retrieves native interface name from user input name, if the alias mode is enabled. User input name can be alias or native interface. API will provide native interface name if the user input is alias name, otherwise it will return the native interface name passed to the API. If the alias mode is not enabled, API will return the input string passed.
+
+   **GetAliasNameFromUIName**  
+   `func GetAliasNameFromUIName(ifName *string) *string`  
+
+   Retrieves alias interface name from user input name, if the alias mode is enabled. User input name can be alias or native interface. API will provide alias name if the user input is native interface name, otherwise it will return the alias name passed to the API. If the alias mode is not enabled, API will return the input string passed.
+
+## 3.2 DB Changes
+
+### 3.2.1 CONFIG DB
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:183]: Where is the "alias" configuration stored?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+   `func GetNativeNameFromUIName(ifName *string) *string`  
+
+   Retrieves native interface name from user input name, if the alias mode is enabled. User input name can be alias or native interface. API will provide native interface name if the user input is alias name, otherwise it will return the native interface name passed to the API. If the alias mode is not enabled, API will return the input string passed.
+
+   **GetAliasNameFromUIName**  
+   `func GetAliasNameFromUIName(ifName *string) *string`  
+
+   Retrieves alias interface name from user input name, if the alias mode is enabled. User input name can be alias or native interface. API will provide alias name if the user input is native interface name, otherwise it will return the alias name passed to the API. If the alias mode is not enabled, API will return the input string passed.
+
+## 3.2 DB Changes
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:180]: Make clear that DB references to ports are unchanged - these still use the native naming, right?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+
+## 3.5 SAI
+
+## 3.6 User Interface
+
+### 3.6.1 Data Models
+Yang model used for alias mode handling  
+[sonic-device-metadata.yang](https://github.com/project-arlo/sonic-mgmt-framework/blob/master/models/yang/sonic/sonic-device-metadata.yang)  
+
+Supported yang objects and attributes are highlighted in green:
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: YANG not yang - throughout

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 2 above)
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: What about : -
- E1/2
- E 1/2 or e 1/2
- Et 1/2 or et 1/2
- ETH 1/2 or ETH 1/2
- etc

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+```
+sonic(conf-if-Eth1/2)# ip address 2.2.2.2/24
+```
+
+#### 3.6.2.2 Show Commands
+
+#### Display alias mode  
+
+`show alias`
+```
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:266]: Shouldn't this return something (i.e not blank?) - "Alias mode is disabled" like below?

I would go with: -
show port-naming = SONiC | standard

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+                       T  Eth1/5/2
+                       T  Eth1/10
+```
+
+
+
+
+#### 3.6.2.3 Debug Commands
+
+#### 3.6.2.4 IS-CLI Compliance
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:318]: Please state that Cisco has no equivalent command - port naming is always fixed.

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+
+#### 3.6.3.1
+
+##### Following REST operations will be supported
+
+**PATCH, PUT, DELETE and GET**
+
+- `​/sonic-device-metadata:sonic-device-metadata​/DEVICE_METADATA​/DEVICE_METADATA_LIST={name}​/aliasMode`
+
+
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:348]: Can we also see examples of RESTCONF port reference in SONiC and standard mode? How is this changed when referring to ports (broken out and not).

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+# 4 Flow Diagrams
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:349]: No gNMI section?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+
+# 5 Error Handling
+
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:352]: What happens if the user (CLI/script, REST, gNMI) refers to the wrong naming mode? Pls show the error message returned.

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 24]
```diff
@@ -0,0 +1,337 @@
...
+Same as CLI show test, but using gNMI GET request, verify the JSON response.
+
+#### Configuration via REST
+
+Same as CLI configuration test, but using REST request
+
+#### Get configuration via REST
+
+Same as CLI configuration test, but using REST request
+
```
Comment (ben-gale) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:408]: gNMI?

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 11 above)
Comment (joyas-joseph) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: Is it OK to kill all existing session when naming mode is changed?

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 1 above)
Comment (joyas-joseph) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: Being futuristic (for chassis case) and also allows the user to easily identify the mode from the name of the interface.

Comment (hsaly) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Any specific reason for starting from slot1 . Usually all Pizza boxes have their interfaces starting from 0 .
Will be soon supporting Subinterfaces . So how is it going to be ?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 25]
```diff
@@ -0,0 +1,337 @@
...
+#### 3.6.2.1 Configuration Commands
+
+#### Alias mode configuration
+
+`[no] alias enable`
+```
+sonic(config)# alias enable
+sonic(config)# no alias enable
+```
+#### Config commands when alias mode is enabled
```
Comment (hsaly) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Please do mention no reload is needed

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 2 above)
Comment (hsaly) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: What if native mode is used to config in alias mode ?

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 26]
```diff
@@ -0,0 +1,337 @@
...
+- Implement alias related config & show commands.
+- REST & gNMI set/get support for alias mode.
+
+### 1.1.3 Scalability Requirements
+
+### 1.1.4 Warm Boot Requirements
+
+## 1.2 Design Overview
+
+### 1.2.1 Basic Approach
```
Comment (hsaly) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:63]: What about upgrade downgrade considerations>

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 27]
```diff
@@ -0,0 +1,337 @@
...
+
+# 5 Error Handling
+
+# 6 Serviceability and Debug
+
+# 7 Warm Boot Support
+
+# 8 Scalability
+
+# 9 Unit Test and Automation
```
Comment (hsaly) [inline doc/mgmt/SONiC_AliasFeature_HLD.md]: Add a test case for upgrade /downgrade between releases

Comment (joyas-joseph) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: There is no upgrade/downgrade case. The interface name in config_db.json is always in native format.

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 9 above)
Comment (bhavini-gada) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: As discussed offline, we will go with interface-naming {native | standard }

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 11 above)
Comment (ben-gale) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: > Is it OK to kill all existing session when naming mode is changed?

It's not ideal. What's the reason for needing to do this? It may be acceptable if there is no practical alternative, but it would be better to solve.

File: doc/mgmt/SONiC_AliasFeature_HLD.md (hunk 1 above)
Comment (ben-gale) [reply doc/mgmt/SONiC_AliasFeature_HLD.md]: Yes I understand that, but I'm asking to explain this in the doc. That is, explain that slot will likely always be 1 in a fixed pizza-box format, but may take different values in other formats (e.g. chassis).

File: doc/mgmt/SONiC_AliasFeature_HLD.md [hunk 28]
```diff
@@ -0,0 +1,411 @@
...
+
+## 2.2 Functional Description
+
+# 3 Design
+
+## 3.1 Overview
+
+- For dynamically using the right string for ethernet interfaces (i.e Ethernet 4 in native mode & Eth 1/4 in standard mode), XML file has to be modified to use the correct PTYPE. `PHY_INTERFACE` has to be used as PTYPE for Ethernet Interface.
+- When interface-naming is set to standard, ethernet interfaces will be represented as Eth[slot/port/breakout-port], where slot will start from 1 and port/breakout-port numbers will start from 1. Slot will always likely to be 1 in fixed pizza-box format, but may take different values in chassis format. This name format is fixed and will not have references to interface speed. For e.g Eth1/1 (in non breakout mode), Eth1/1/1 (in breakout mode).
+- The standard interface name {e.g Eth1/1) is picked from platform.json file. All platforms need to be updated with this format for port name.
```
Comment (vishnushetty) [inline doc/mgmt/SONiC_AliasFeature_HLD.md:85]: Please confirm alias  is picked up from CONFIG_DB not from platform.json, the back end application populate DB entries from platform.json based on break-out mode.

Comment (bhavini-gada) [reply doc/mgmt/SONiC_AliasFeature_HLD.md:85]: Need updated platform.json with the new format for ports. Interface-naming API's use data from CONFIG_DB.
